### ISSUE TITLE
Lag when sending commands for switch/number/select

### ISSUE THREAD

**Issue by alexdelprete** [alexdelprete/HA-NeoPool-MQTT#3]
(cross-posted from tasmota repo [discussion](https://github.com/arendst/Tasmota/discussions/19811#discussioncomment-7414338))

@curzon01 I found a major problem for the MQTT switch/number/select entities (and possibly, all modifiable/actionable MQTT entities, well described in HA documentation:

![image](https://github.com/arendst/Tasmota/assets/7027842/67ec09ae-ac8c-489f-aa2d-606695847ddc)

This is the configuration of a MQTT Switch:

```yaml
  switch:
    - unique_id: "neopool_mqtt_light"
      name: "NeoPool MQTT Light"
      device_class: switch
      command_topic: "cmnd/SmartPool/NPLight"
      state_topic: "tele/SmartPool/SENSOR"
      retain: true
      optimistic: false
      value_template: >-
        {{ value_json.NeoPool.Light }}
      payload_on: 1
      payload_off: 0
      state_on: 1
      state_off: 0
      availability_topic: "tele/SmartPool/LWT"
      payload_available: "Online"
      payload_not_available: "Offline"
```

When you start HA, or reload the config, these entities are reinitialized. I'm using `tele/SmartPool/SENSOR` to determine the state of the switch. Problem is that if the SENSOR topic is not retained, all the entities are unavailable until the TelePeriod timer expires and the data is refreshed. So in the readme I'm recommending a configuration with `TelePeriod 10` and `SensorRetain 1`. With this config, when you restart/reload HA the NeoPool entities are immediately available. The only minor issue is that if you change the switch state by clicking it, HA sends the command, but it has to wait another TelePeriod cycle to reflect the state change of the switch.

One solution (cleanest one) is to use `SO4 1` and `state_topic`: `stat/SmartPool/NPLIGHT`, that contains the immediate result of the command. This works, and it's immediate when you click the switch, but it has one major issue I wasn't able to solve: when you restart/reload HA, the `stat/SmartPool/NPLIGHT` topic is not available, since it doesn't have the retain flag (like with the SensorRetain command, that I'm using so on HA restart/reload the entities are immediately available), and I found no Tasmota command to enable retain on the result of a command.

I already tried the Backlog compound command, but unfortunately it doesn't work. I think because value_template is parsed before sending the commands, so we still wait the TelePeriod cycle when activating the switch. :(

The only real solution would be to have stat/SmartPool/%command% retainable. But that is a Tasmota FR because I didn't find a way to do that in the documentation.

![image](https://github.com/alexdelprete/HA-NeoPool-MQTT/assets/7027842/7a61e6a8-baca-440d-a5a7-3768a723d695)

**Comment by curzon01** [comment]
Sorry, this way the Delay trick is also confusing

An issue is also the retain for commands, this is a no go

I prefere to use "optimistic: true" with "retain false", that's works

```yaml
    - unique_id: "neopool_mqtt_light"
      name: "NeoPool MQTT Light"
      device_class: switch
      command_topic: "cmnd/SmartPool/NPLight"
      state_topic: "tele/SmartPool/SENSOR"
      retain: false
      optimistic: true
      value_template: >-
        {{ value_json.NeoPool.Light }}
      payload_on: "NPLight 1;Delay 25;Teleperiod"
      payload_off: "NPLight 0;Delay 25;Teleperiod"
      state_on: 1
      state_off: 0
      availability_topic: "tele/SmartPool/LWT"
      payload_available: "Online"
      payload_not_available: "Offline"
```

**Comment by alexdelprete** [comment]
the retain: true was there because I thought about using cmnd/%command% to get the status immediately, but since it's not a response it wouldn't work because the state would be inconsistent if you changed state using the display or another command. I will revert to retain: false for commands.

I evaluated `optimistic: true`, but it's a workaround really. I don't like it particularly. But if there's no way to set retain flag on stat/%commands% I guess it's a possible solution.

If you use optimistic, there's no need to use Delay. Furthermore, Delay 25 is like using TelePeriod 30, so you click and wait 25-30s for the response. It's tha same issue. :)

There's no good workaround...I'll tell users to simply wait...:)

**Comment by alexdelprete** [comment]
Testing NPLight sensor with these configurations:

```yaml
TelePeriod back to 30s and sensor config:

      state_topic: "stat/SmartPool/NPLight"
      optimistic: false
```

At startup (I reload HA config), since the topic does not exist (no command ever received and not retained), the status of the sensor is `unknown`, and it makes sense.

![image](https://github.com/alexdelprete/HA-NeoPool-MQTT/assets/7027842/5a01aed5-8b9e-4357-9499-af7e1656f65f)

Then I tested the exact same config, but with optimistic on:

```yaml
TelePeriod back to 30s and sensor config:

      state_topic: "stat/SmartPool/NPLight"
      optimistic: true
```

Now I reload the config and....

![image](https://github.com/alexdelprete/HA-NeoPool-MQTT/assets/7027842/9447c369-8f30-44a5-9ecc-f7c58d599af3)

Good, it's not unknown, but...

![image](https://github.com/alexdelprete/HA-NeoPool-MQTT/assets/7027842/3b38dbff-b922-4f91-8818-d3f88c0bb463)

The real status is really On. That's why I don't like Optimistic. It's a dirty workaround...

There's no real clean solution, except for stat/%commands% being retained.

Since I prefer state consistency over reaction times of a command, I'll write a note telling the users that when they change a config or click a switch they need to wait `TelePeriod` seconds.

If I open a FR on tasmota for having a command that sets stat/%commands% retain flag, would it be considered?

**Comment by curzon01** [comment]
> I will revert to retain: false for commands

thx, retained commands for devices in MQTT are the pure catastrope

> Delay 25 is like using TelePeriod 30, so you click and wait 25-30s for the response. It's the same issue. :)

Delay param are 1/10 sec so Delay 25 are 2.5 sec, not 25

**Comment by alexdelprete** [comment]
> Delay param are 1/10 sec so Delay 25 are 2.5 sec, not 25

Ok. But the above code is wrong for `command_topic`, it should send Backlog command, not NPLight, right? I'll test it.

> I prefere to use "optimistic: true" with "retain false", that's works
> 
> ```yaml
>     - unique_id: "neopool_mqtt_light"
>       name: "NeoPool MQTT Light"
>       device_class: switch
>       command_topic: "cmnd/SmartPool/NPLight"
>       state_topic: "tele/SmartPool/SENSOR"
>       retain: false
>       optimistic: true
>       value_template: >-
>         {{ value_json.NeoPool.Light }}
>       payload_on: "NPLight 1;Delay 25;Teleperiod"
>       payload_off: "NPLight 0;Delay 25;Teleperiod"
>       state_on: 1
>       state_off: 0
>       availability_topic: "tele/SmartPool/LWT"
>       payload_available: "Online"
>       payload_not_available: "Offline"
> ```

**Comment by alexdelprete** [comment]
Tested it, it works with Delay 30. With 25 it was not consistent.

When 2 days ago I tested the backlog trick, without delay, it wasn't working. Why does it need a 3s delay?

Question remains though: the `stat/` topic should contain the status. We're getting status from telemetry, and that's why the implementation is not really efficient.

**Comment by curzon01** [comment]
>  Why does it need a 3s delay?

i don't know, i guess this is related to the internal processing of mqtt topics of HA, the logic doesn't make sense to me either, because this should all be independent.

> the stat/ topic should contain the status

I also think so, `SO4 1` and listen on stat/xxxx/<cmnd>, what you are missing is that these are retained

**Comment by alexdelprete** [comment]
> I also think so, `SO4 1` and listen on `stat/xxxx/<cmnd>`, what you are missing is that these are retained

Yes. And if those were retained, updates would be immediate and consistent and on restart the status wouldn't be `unknown`.

Is that something that could be implemented at tasmota level as with `SensorRetain`?

Ideally, something like `StatCmdRetain 1` that would automatically do `SO4 1` and also set the retain flag for the relevant stat/%command% topics.

Would be perfect.

**Comment by alexdelprete** [comment]
Anyway, I updated all configurable entities (switches/selects/numbers) with the backlog/delay logic. It was trickier than expected because of the way HA manages templates (everything is a string):

```yaml
      command_template: >-
        {% set values = { 'Manual':0, 'Auto':1, 'Heating':2, 'Smart':3, 'Intelligent':4, 'Backwash':13 } %}
        {% set setvalue = values[value] | string if value in values.keys() %}
        {{ 'NPFiltrationmode ' + setvalue +';Delay 30;Teleperiod' }}
```

**Comment by curzon01** [comment]
the notation like 

```yaml
      command_template: >-
        {% set values = { 'Manual':0, 'Auto':1, 'Heating':2, 'Smart':3, 'Intelligent':4, 'Backwash':13 } %}
        {% set setvalue = values[value] | string if value in values.keys() %}
        {{ 'NPFiltrationmode ' + setvalue +';Delay 30;Teleperiod' }}
```

not necessary with yaml, because that is one of the advantages of yaml and at the same time its biggest disadvantage

you can write simply as it was before

```yaml
      command_template: >-
        {% set values = { 'Manual':'0', 'Auto':'1', 'Heating':'2', 'Smart':'3', 'Intelligent':'4', 'Backwash':'13' } %}
        NPFiltrationMode {{ values[value] if value in values.keys() }};Delay 30;Teleperiod
```

**Comment by alexdelprete** [comment]
I know I can, but if in future I need to modify the code of setvalue, I prefer to keep it separate. It's more a personal thing based on my experience when I started with HA and yaml templates, and I created one of the most complex and huge automations existing for my Nuki Lock (I didn't know how to create python custom components, so I used HA standard set of functionalities/services): https://community.home-assistant.io/t/nuki-card-with-callback-support-supports-both-lock-opener-it-replaces-the-official-integration/311932

That's how I managed to learn how HA works and all the errors to avoid for maintenance of code.

In this case it's probably overkill, but whenever I work with HA stuff, I always use the same methodology based on my personal experience. :)

**Comment by curzon01** [comment]
To get the status immediately, there is the idea to use SO4 1 and set the status topic to state/xxxx/<command_name>. For this, according to your idea, there should be a new Tasmota retain option for SO4 1 topics.

I had thought about your suggestion yesterday, it is implemented quite fast (I used the command "ResponseRetain"), but it can only be implemented incompletely and such a PR would probably not be accepted:
The Tasmota philosophy for retain options for topics is, if you turn them on, the corresponding topics are treated retain. If you turn them off, however, Tasmota deletes the corresponding topics. This is also an absolutely necessary behavior and this is not a big problem for the existing options (ButtonRetain, InfoRetain, PowerRetain, SensorRetain, StateRetain....).

To keep this Tasmota would have to remember all sent commands when ResponseRetain is active, in order to remove them when ResponseRetain is OFF, and the memory of the small ESPs is not enough for this. This is also the reason why there is no such option.

**Comment by curzon01** [comment]
It could also solved on HA side:

- You use SO4 1 and the command topics as state
- A (Python) script gets the current state at HA start (Event) via TelePeriod and sets the HA entities accordingly.

Would this be feasible?

**Comment by curzon01** [comment]
That being said, to really fully implement the condition, HA's behavior using SO4 1 will not be sufficient, as I am annoyed with this behavior in other implementations like my Panasonic heat pump using Heishamon.

Example: Manual filtration is switched by third party (not via HA) and changes the status of the Machine. For example, it goes from "Smart" to "Manual" but HA does not notice this. The mode switch is still set to "Smart". To be able to switch the thing to "Smart" back, I must first select another mode (for HA I cannot select "Smart" twice) and then switch back to "Smart".

Therefore, the evaluation of a topic of the overall state of the device is actually the more correct and that would be SENSOR, right?

**Comment by alexdelprete** [comment]
> Therefore, the evaluation of a topic of the overall state of the device is actually the more correct and that would be SENSOR, right?

Maybe you missed my [comment](https://github.com/alexdelprete/HA-NeoPool-MQTT/issues/3#issuecomment-1787455345) of 2 days ago: I reached that same conclusion.

![image](https://github.com/alexdelprete/HA-NeoPool-MQTT/assets/7027842/b89efc34-5aca-40bf-859e-2e7958350045)

So the problem is that if the real-time state of NeoPool entities (independent of who controls the entities, the hw, HA, tasmota UI, etc.) is ONLY in SENSOR, we're subject to TelePeriod for updates. There's no alternative solution, unless we come up with a new design.

From what I see, ALL other tasmota projects rely on retained entities to have the state in real-time. I don't see any workaround to that. So the only idea is that the NeoPool driver creates and maintains these retained entities, that means new topics.

Anyhow, right now all the switches (I released a new version with AUX switches yesterday and updated UI, PC and mobile) react in 3s when activating them and all other entities/statuses in 20s (with TelePeriod 20). We can't do better than this with current design of things.

I think Tasmota should tackle this challenge...I feel its design is still to much entrenched in the "switch" device use-case, but Tasmota is now used for many more things, it should take into account many other use-cases.

**Comment by curzon01** [comment]
> So the only idea is that the NeoPool driver creates and maintains these retained entities, that means new topics

agree, I had the same in mind, will think about it....

**Comment by curzon01** [comment]
see #9

**Comment by Karel311** [comment]
Hi Alex,

I have set up a atom lite+tail485 configuration to replace an ESP32 running ESPhome on my oxilife

I have noticed in home assistant that binary sensors or states (filtration for example) receive an unavailable status for a few seconds right before updating their values.  It happens random (2 to 10 times per hour)

no faults or restarts in the tasmota logs - only temporary unvailable states in home assistant for 1 to 10 seconds. 
wifi is working fine - no signs of disconnection there (access point 5 meters from the atom lite)

is there a way to retain the value till the next send value?
Or is it related with the issue discussed above?

Best regards and really nice job on this project!
k
